src/source-updater.js

```diff
@@ -24,54 +24,136 @@ const actions = {
   },
   callback: (callback) => (type, sourceUpdater) => {
     callback();
+  },
+  duration: (duration) => (sourceUpdater) => {
+    sourceUpdater.mediaSource.duration = duration;
   }
 };
 
 const updating = (type, sourceUpdater) => {
   const sourceBuffer = sourceUpdater[`${type}Buffer`];
 
-  return (sourceBuffer && sourceBuffer.updating) || sourceUpdater.queue[type].pending;
+  return (sourceBuffer && sourceBuffer.updating) || sourceUpdater.queuePending[type];
 };
 
-const shiftQueue = (type, sourceUpdater) => {
-  const queue = sourceUpdater.queue[type];
-
-  if (updating(type, sourceUpdater) || !queue.actions.length || !sourceUpdater.started_) {
-    return;
+const processMediaSourceQueueEntry = (index, sourceUpdater) => {
+  if (updating('audio', sourceUpdater) || updating('video', sourceUpdater)) {
+    // block until all buffers are ready for use
+    return false;
   }
 
-  const action = queue.actions.shift();
+  const queueEntry = sourceUpdater.queue.splice(index, 1)[0];
+
+  queueEntry.action(sourceUpdater);
```

Comment:
is `sourceUpdater` the right argument for this? Looks like we expect duration:
```
    pushQueue({
      type: 'mediaSource',
      sourceUpdater: this,
      action: actions.duration(duration),
      name: 'duration',
      doneFn
    });
```

Reply:
The action is a function that returns a function. The returned function expects the `sourceUpdater`.

Reply:
Ah, you're right; forgot that duration was passed in from setDuration